schedSetter/unix/schedSetter.c

```diff
@@ -0,0 +1,258 @@
+
+// The schedSetter program sets scheduling priority of an executed process to
+// realtime, then switches it to and fro on user input on a given FIFO special
+// file (see below). The first is meant to give isolation between a realtime
+// process and the privileged process capable of setting realtime scheduling
+// policies, preventing a remote root exploit following a remote code
+// execution. The second is just a convenience meant for revoking real-time
+// scheduling from a non-malicious process - some engineering using seccomp-bpf
+// (on Linux) would be necessary to enable us to reliably take back realtime
+// scheduling from the executed process, if it is malicious.
+//
+// Invoke with root privilege and four arguments:
+// schedSetter fifo uid gid command
+// For example; if sudo gives root rights, stdinFIFO is an appropriate FIFO
+// special file, "1000" are an appropriate user's UID and GID, and the target
+// executable is called scantool; it could be called with:
+// sudo schedSetter stdinFIFO 1000 1000 scantool
+//
+// See schedSetterWrapper.sh.
+//
+// The scheduling policy switching function of schedSetter is controlled by
+// writing to the FIFO special file, any input triggers a switch; so if the
+// file is in the usual line blocking mode, you should just press enter; in
+// non-blocking mode any character should work.
+//
+// Only tested on Linux.
+
+// TODO: would it make sense to also set the Linux ioprio stuff?
+
+#define _BSD_SOURCE     // for setgroups on Linux
+#define _DEFAULT_SOURCE // Glibc ...
+#include <grp.h>
+#undef _DEFAULT_SOURCE
+#undef _BSD_SOURCE
+
+#include <ctype.h>  // isdigit
+#include <stdio.h>  // fgetc
+#include <stdlib.h> // atol
+#include <string.h> // memcpy
+
+#include <sched.h>
+#include <sys/mman.h> // mlockall
+#include <unistd.h>   // fork, exec
+
+//#include "debugging.h"
+
+// Set given uid and gid, drop supplementary groups. If uid is zero, use getuid
+// and getgid.
+static int
+dropPrivileges(uid_t uid, gid_t gid) {
+	/*
+	if (setuid(0) < 0) {
+	}
+	*/
+
+	if (uid == 0) {
+		uid = getuid();
+		gid = getgid();
+	}
+	if (setgroups(0, NULL) < 0) {
+		return -19;
+	}
+	if (setgid(gid) < 0) {
+		return -20;
+	}
+	if (setuid(uid) < 0) {
+		return -21;
+	}
+
+	/*
+	// Debugging
+	if (setuid(0) < 0) {
+	}
+	*/
+
+	return 0;
+}
+
+static int
+f(char inputFile[], uid_t uid, gid_t gid, char targetCommand[]) {
+	int oldPolicy, policy = SCHED_FIFO;
+	struct sched_param oldParam, param;
+	int min, max;
+
+	int pid;
+
+	int returnValue;
+
+	FILE *inputF;
+
+	// Fork and execute target process.
+	pid = fork();
+	if (pid < 0) {
+		return 3;
+	}
+	if (pid == 0) {
+		/*
+		// Debugging
+		if (printUserGroup() < 0) {
+			return 30;
+		}
+		*/
+
+		if (dropPrivileges(uid, gid) < 0) {
+			return 31;
+		}
+
+		/*
+		// Debugging
+		if (printUserGroup() < 0) {
+			return 32;
+		}
+		*/
+
+		execlp(targetCommand, targetCommand, (char *)NULL);
+		return 4;
+	}
+
+	// Get valid priority ranges.
```

Comment:
SchedSetter gets valid priorities range using sched_get_priority_ functions.

Reply:
I made some changes to the commit; none changes to the program, just changed comments and the like.